### PR TITLE
add FTPS support for vdisk backups

### LINKED ISSUE #414: Look into FTPS support for backups
Currently only plain FTP is supported for vdisk backup purposes. Look into supporting FTPS (FTP over TLS) as well.

### PR DESCRIPTION
**Pull request by GlenDC**
Closes #414.

### CODE REVIEW

File: zeroctl/cmd/backup/export.go [hunk 1]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+		"tls-insecure", "i", false,
+		"when given, TLS accepts any certificate")
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.ServerName,
+		"tls-server", "n", "",
+		"when given, TLS verifies the hostname")
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.Certificate,
+		"tls-cert", "e", "",
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: `-e` doesn't make sense at all. Please don't support shorthand one-char flags for the tls-flags as they simply are not clear enough. Use `StringVar` instead of `StringVarP`, and thus only support `--tls-cert`.

File: zeroctl/cmd/backup/export.go [hunk 2]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+		"a sertificate, provided by the client")
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.Key,
+		"tls-key", "t", "",
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: `-t` doesn't make sense at all. Please don't support shorthand one-char flags for the tls-flags as they simply are not clear enough. Use `StringVar` instead of `StringVarP`, and thus only support `--tls-key`.

File: zeroctl/cmd/backup/export.go [hunk 3]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+		"a private key, provided by the client")
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.CA,
+		"tls-ca", "a", "",
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: `-a` doesn't make sense at all. Please don't support shorthand one-char flags for the tls-flags as they simply are not clear enough. Use `StringVar` instead of `StringVarP`, and thus only support `--tls-ca`.

File: zeroctl/cmd/backup/export.go [hunk 4]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
 		&vdiskCmdCfg.Force,
 		"force", "f", false,
 		"when given, overwrite a deduped map if it can't be loaded")
+
+	ExportVdiskCmd.Flags().BoolVarP(
+		&vdiskCmdCfg.TLSConfig.InsecureSkipVerify,
+		"tls-insecure", "i", false,
+		"when given, TLS accepts any certificate")
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: change description to:

> when given FTP over SSL will be used without cert verification

File: zeroctl/cmd/backup/export.go [hunk 5]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.ServerName,
+		"tls-server", "n", "",
+		"when given, TLS verifies the hostname")
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: change description to:

> certs will be verified when given (required when --tls-insecure is not used)`

File: zeroctl/cmd/backup/export.go [hunk 6]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.Certificate,
+		"tls-cert", "e", "",
+		"a sertificate, provided by the client")
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.Key,
+		"tls-key", "t", "",
+		"a private key, provided by the client")
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: change description to:

> PEM-encoded file containing the private TLS client key

File: zeroctl/cmd/backup/export.go [hunk 7]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+		"when given, TLS accepts any certificate")
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.ServerName,
+		"tls-server", "n", "",
+		"when given, TLS verifies the hostname")
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.Certificate,
+		"tls-cert", "e", "",
+		"a sertificate, provided by the client")
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: change description to:

> PEM-encoded file containing the TLS Client cert (FTPS will be used when given)

File: zeroctl/cmd/backup/export.go [hunk 8]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.Key,
+		"tls-key", "t", "",
+		"a private key, provided by the client")
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.CA,
+		"tls-ca", "a", "",
+		"path to the certificate pool")
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: change description to:

> optional PEM-encoded file containing the TLS CA Pool (defaults to system pool when not given)

File: zeroctl/cmd/backup/export.go [hunk 9]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
 		&vdiskCmdCfg.Force,
 		"force", "f", false,
 		"when given, overwrite a deduped map if it can't be loaded")
+
+	ExportVdiskCmd.Flags().BoolVarP(
+		&vdiskCmdCfg.TLSConfig.InsecureSkipVerify,
+		"tls-insecure", "i", false,
```
[the remark this answers is not in the document]
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 4 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go [hunk 10]
```diff
@@ -147,4 +164,27 @@ using a different private key or compression type, than the one(s) used right no
...
+		"when given, TLS accepts any certificate")
+
+	ExportVdiskCmd.Flags().StringVarP(
+		&vdiskCmdCfg.TLSConfig.ServerName,
+		"tls-server", "n", "",
```
[the remark this answers is not in the document]
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 5 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 1 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 7 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 2 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 6 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 3 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go (hunk 8 above)
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: This comment also applies to the `import vdisk` and `describe snapshot` commands

File: zeroctl/cmd/backup/export.go [hunk 11]
```diff
@@ -45,17 +45,34 @@ func exportVdisk(cmd *cobra.Command, args []string) error {
...
+		BlockSize:          exportVdiskCmdCfg.ExportBlockSize,
+		BlockStorageConfig: vdiskCmdCfg.SourceConfig,
+		JobCount:           vdiskCmdCfg.JobCount,
+		CompressionType:    vdiskCmdCfg.CompressionType,
+		CryptoKey:          vdiskCmdCfg.PrivateKey,
+		Force:              vdiskCmdCfg.Force,
 	}
 
+	var tlsConfig *backup.TLSClientConfig
+	if vdiskCmdCfg.TLSConfig.InsecureSkipVerify || vdiskCmdCfg.TLSConfig.Certificate != "" ||
```
[the remark this answers is not in the document]
Comment (katia-e) [reply zeroctl/cmd/backup/export.go]: Shall there also be server name check?
```go
if vdiskCmdCfg.TLSConfig.InsecureSkipVerify || vdiskCmdCfg.TLSConfig.CertFile != "" ||
   vdiskCmdCfg.TLSConfig.CAFile != ""  ||  vdiskCmdCfg.TLSConfig.ServerName !="" {
```

In case if only --tls-serve flag is given, TLS config will verify the host

[the remark this answers is not in the document]
Comment (GlenDC) [reply zeroctl/cmd/backup/export.go]: Ah OK, leave it in, in that case :)

File: docs/zeroctl/commands/describe.md [hunk 12]
```diff
@@ -24,6 +24,12 @@ Alternatively you can also give a local directory path to the `--storage` flag,
 to backup to the local file system instead.
 This is also the default in case the `--storage` flag is not specified.
 
+When both the `--storage` and at least one of
```
Comment (GlenDC) [inline docs/zeroctl/commands/describe.md]: The first part should probably become:

> When the `--storage` flag contains an FTP storage config and at least one of

File: docs/zeroctl/commands/describe.md [hunk 13]
```diff
@@ -70,6 +81,12 @@ $ zerodisk describe snapshot foo -s ftp://1.2.3.4:21 --pretty
 Just as with the [export][export] and [import][import] commands,
 can an encrypted snapshot be described by specifying the `-k`/`--key` (private key) flag.
 
+If we want to connect over FTPS we can add necessary `tls` flags:
+
+```
+$ describe snapshot foo -s ftp://1.2.3.4:21 --tls-server 1.2.3.4 --tls-cert sample.sert --tls-key sample.key
```
Comment (GlenDC) [inline docs/zeroctl/commands/describe.md]: use a backslash to keep the lines short and don't forget the `zeroctl` part in the front:

```
$ zeroctl describe snapshot foo -s ftp://1.2.3.4:21 \
    --tls-server 1.2.3.4 \
    --tls-cert sample.sert --tls-key sample.key
```

File: docs/zeroctl/commands/export.md [hunk 14]
```diff
@@ -42,6 +42,13 @@ using a different private key or compression type, than the one(s) used right no
 By default LZ4 compression is used, which is the fastest of the supported compression algorithms.
 XZ compression can be used, which has a better compression ratio but slows down the export of a vdisk.
 
+When both the `--storage` and at least one of
```
Comment (GlenDC) [inline docs/zeroctl/commands/export.md]: The first part should probably become:

> When the `--storage` flag contains an FTP storage config and at least one of

File: docs/zeroctl/commands/export.md [hunk 15]
```diff
@@ -94,5 +106,11 @@ If we want to create a public backup (without encryption) we can omit the `-k` f
 $ zerodisk export vdisk a -s ftp://1.2.3.4:21 --config 1.2.3.4:2000
 ```
 
+If we want to connect over FTPS we can add necessary `tls` flags:
+
+```
+$ zerodisk export vdisk a -s ftp://1.2.3.4:21 --tls-server 1.2.3.4 --tls-cert sample.sert --tls-key sample.key
```
Comment (GlenDC) [inline docs/zeroctl/commands/export.md]: use backslash to keep the line shorter (~60 chars per line if possible)

File: docs/zeroctl/commands/import.md [hunk 16]
```diff
@@ -32,6 +32,13 @@ Alternatively you can also give a local directory path to the `--storage` flag,
 to backup to the local file system instead.
 This is also the default in case the `--storage` flag is not specified.
 
+When both the `--storage` and at least one of
```
Comment (GlenDC) [inline docs/zeroctl/commands/import.md]: The first part should probably become:

> When the `--storage` flag contains an FTP storage config and at least one of

File: docs/zeroctl/commands/import.md [hunk 17]
```diff
@@ -88,5 +100,11 @@ If we want to import a public backup we can simply omit the `-k` flag as we no l
 $ zerodisk import vdisk a mybackup -s ftp://1.2.3.4:21 --config 1.2.3.4:2000
 ```
 
+If we want to connect over FTPS we can add necessary `tls` flags:
+
+```
+$ zerodisk import vdisk a mybackup -s ftp://1.2.3.4:21 --tls-server 1.2.3.4 --tls-cert sample.sert --tls-key sample.key
```
Comment (GlenDC) [inline docs/zeroctl/commands/import.md]: Use backslash to keep the lines of this example shorter (~60 chars if possible)

File: docs/zeroctl/commands/import.md [hunk 18]
```diff
@@ -88,5 +100,11 @@ If we want to import a public backup we can simply omit the `-k` flag as we no l
 $ zerodisk import vdisk a mybackup -s ftp://1.2.3.4:21 --config 1.2.3.4:2000
 ```
 
+If we want to connect over FTPS we can add necessary `tls` flags:
```
Comment (GlenDC) [inline docs/zeroctl/commands/import.md]: Change `tls` with TLS

File: docs/zeroctl/commands/export.md [hunk 19]
```diff
@@ -94,5 +106,11 @@ If we want to create a public backup (without encryption) we can omit the `-k` f
 $ zerodisk export vdisk a -s ftp://1.2.3.4:21 --config 1.2.3.4:2000
 ```
 
+If we want to connect over FTPS we can add necessary `tls` flags:
```
Comment (GlenDC) [inline docs/zeroctl/commands/export.md]: Change `tls` with TLS

File: docs/zeroctl/commands/describe.md [hunk 20]
```diff
@@ -70,6 +81,12 @@ $ zerodisk describe snapshot foo -s ftp://1.2.3.4:21 --pretty
 Just as with the [export][export] and [import][import] commands,
 can an encrypted snapshot be described by specifying the `-k`/`--key` (private key) flag.
 
+If we want to connect over FTPS we can add necessary `tls` flags:
```
Comment (GlenDC) [inline docs/zeroctl/commands/describe.md]: Change `tls` with TLS

File: nbd/ardb/backup/storage_ftp.go [hunk 21]
```diff
@@ -126,23 +185,50 @@ func (cfg *FTPStorageConfig) validate() error {
...
 
-	address := strings.TrimPrefix(cfg.Address, "ftp://")
+	// if a tls config is specified,
+	// create an actual TLS Config, and pass it onto the goftp Config
+	if cfg.TLSConfig != nil {
+		config.TLSConfig, err = tlsClientAuth(*cfg.TLSConfig)
+		if err != nil {
+			return nil, err
+		}
+		err = cfg.TLSConfig.validate()
```
Comment (GlenDC) [inline nbd/ardb/backup/storage_ftp.go]: This validation should happen before you do `config.TLSConfig, err = tlsClientAuth(*cfg.TLSConfig)`, otherwise what's the point...

File: zeroctl/cmd/backup/backup.go [hunk 22]
```diff
@@ -1,19 +1,153 @@
...
-	CompressionType     backup.CompressionType // optional
-	JobCount            int                    // optional
-	Force               bool                   //optional
+	VdiskID         string                 // required
+	SourceConfig    config.SourceConfig    // optional
+	SnapshotID      string                 // optional
+	PrivateKey      backup.CryptoKey       // optional
+	CompressionType backup.CompressionType // optional
+	JobCount        int                    // optional
+	Force           bool                   //optional
```
Comment (GlenDC) [inline zeroctl/cmd/backup/backup.go]: not your fault, but while you're at it could you add a missing space here (between `//` and `optional`)

File: zeroctl/cmd/backup/describe.go [hunk 23]
```diff
@@ -39,16 +39,37 @@ func describeSnapshot(cmd *cobra.Command, args []string) error {
...
-	// create (backup) storage driver (so we can read snapshot header from it)
-	driver, err := backup.NewStorageDriver(vdiskCmdCfg.BackupStorageConfig)
-	if err != nil {
-		return err
+	var storageDriver backup.StorageDriver
+
+	if vdiskCmdCfg.BackupStorageConfig.StorageType == ftpStorageType {
+		var ftpStorageConfig backup.FTPStorageDriverConfig
+		ftpStorageConfig.ServerConfig = vdiskCmdCfg.BackupStorageConfig.Resource.(backup.FTPServerConfig)
+		if vdiskCmdCfg.TLSConfig.InsecureSkipVerify || vdiskCmdCfg.TLSConfig.CertFile != "" ||
```
Comment (GlenDC) [inline zeroctl/cmd/backup/describe.go]: you can remove the `vdiskCmdCfg.TLSConfig.ServerName` check, or else update the documentation

File: zeroctl/cmd/backup/describe.go [hunk 24]
```diff
@@ -136,6 +157,13 @@ here are some examples of valid values for that flag:
 Alternatively you can also give a local directory path to the --storage flag,
 to backup to the local file system instead.
 This is also the default in case the --storage flag is not specified.
+
+  When both the --storage and at least one of
```
Comment (GlenDC) [inline zeroctl/cmd/backup/describe.go]: The first part should probably become:

> When the `--storage` flag contains an FTP storage config and at least one of

File: zeroctl/cmd/backup/export.go [hunk 25]
```diff
@@ -45,17 +45,31 @@ func exportVdisk(cmd *cobra.Command, args []string) error {
...
+		CryptoKey:          vdiskCmdCfg.PrivateKey,
+		Force:              vdiskCmdCfg.Force,
 	}
 
+	if vdiskCmdCfg.BackupStorageConfig.StorageType == ftpStorageType {
+		storageConfig := backup.FTPStorageDriverConfig{
+			ServerConfig: vdiskCmdCfg.BackupStorageConfig.Resource.(backup.FTPServerConfig),
+		}
+		if vdiskCmdCfg.TLSConfig.InsecureSkipVerify || vdiskCmdCfg.TLSConfig.CertFile != "" ||
+			vdiskCmdCfg.TLSConfig.CAFile != "" || vdiskCmdCfg.TLSConfig.ServerName != "" {
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: you can remove the `vdiskCmdCfg.TLSConfig.ServerName != ""` check, or else update the docs

File: zeroctl/cmd/backup/export.go [hunk 26]
```diff
@@ -119,7 +133,14 @@ This is also the default in case the --storage flag is not specified.
   When the --force flag is given,
 a deduped map will be overwritten if it already existed,
 AND if it couldn't be loaded, due to being corrupt or encrypted/compressed,
-using a different private key or compression type, than the one(s) used right now.
+using a different private key or compression type, than the one(s) used right now.s
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: Undo this change. I guess you made a typo while saving...

File: zeroctl/cmd/backup/export.go [hunk 27]
```diff
@@ -119,7 +133,14 @@ This is also the default in case the --storage flag is not specified.
...
+
+  When both the --storage and at least one of
```
Comment (GlenDC) [inline zeroctl/cmd/backup/export.go]: The first part should probably become:

> When the `--storage` flag contains an FTP storage config and at least one of

File: zeroctl/cmd/backup/import.go [hunk 28]
```diff
@@ -63,16 +63,30 @@ func importVdisk(cmd *cobra.Command, args []string) error {
...
+		CryptoKey:          vdiskCmdCfg.PrivateKey,
+		Force:              vdiskCmdCfg.Force,
 	}
 
+	if vdiskCmdCfg.BackupStorageConfig.StorageType == ftpStorageType {
+		storageConfig := backup.FTPStorageDriverConfig{
+			ServerConfig: vdiskCmdCfg.BackupStorageConfig.Resource.(backup.FTPServerConfig),
+		}
+		if vdiskCmdCfg.TLSConfig.InsecureSkipVerify || vdiskCmdCfg.TLSConfig.CertFile != "" ||
+			vdiskCmdCfg.TLSConfig.CAFile != "" || vdiskCmdCfg.TLSConfig.ServerName != "" {
```
Comment (GlenDC) [inline zeroctl/cmd/backup/import.go]: Remove this `vdiskCmdCfg.TLSConfig.ServerName != ""` check, or else update the docs

File: zeroctl/cmd/backup/import.go [hunk 29]
```diff
@@ -249,6 +263,14 @@ here are some examples of valid values for that flag:
 Alternatively you can also give a local directory path to the --storage flag,
 to backup to the local file system instead.
 This is also the default in case the --storage flag is not specified.
+
+
+When both the --storage and at least one of
```
Comment (GlenDC) [inline zeroctl/cmd/backup/import.go]: The first part should probably become:

> When the `--storage` flag contains an FTP storage config and at least one of

**Comment by GlenDC** [comment]
LGTM as it is! 👍 

@katia-e are you OK with the code as it is? And are you sure you tested importing/exporting/describing backups using a secure FTPS server?

**Comment by katia-e** [comment]
TLS flags --tls-server, --tls-insecure, --tls-certificate, --tls-key are added to enable FTPS support for backup with zeroctl commands.
TLS connection was tested for `zeroctl import vdisk`,  `zeroctl export vdisk` and `zeroctl describe snapshot` for the various server configurations both with and without client authentication.

All test were successful, the branch is ready for merging.
